crates/ursa-rpc-service/src/api.rs

```diff
@@ -79,65 +91,188 @@ pub trait NetworkInterface: Sync + Send + 'static {
     pub network_send: Sender<NetworkCommand>,
     pub provider_send: Sender<ProviderCommand>,
+    pending_requests: PendingRequests,
+    client: Arc<Client>,
+    origin_config: OriginConfig,
+}
+
+impl<S> NodeNetworkInterface<S>
+where
+    S: Blockstore + Store + Send + Sync + 'static,
+{
+    pub fn new(
+        store: Arc<UrsaStore<S>>,
+        network_send: Sender<NetworkCommand>,
+        provider_send: Sender<ProviderCommand>,
+        origin_config: OriginConfig,
+    ) -> Self {
+        Self {
+            store,
+            network_send,
+            provider_send,
+            origin_config,
+            pending_requests: Arc::new(RwLock::new(HashMap::new())),
+            client: Arc::new(Client::new()),
+        }
+    }
+
+    /// Fetch content from the network
+    async fn get_network(&self, root_cid: Cid) -> Result<()> {
+        info!("Fetching cid {root_cid} from network");
+        let (send, recv) = oneshot::channel();
+
+        // Dispatch bloom filtered bitswap request
+        self.network_send.send(NetworkCommand::GetBitswap {
+            cid: root_cid,
+            sender: send,
+        })?;
+
+        recv.await?
+    }
+
+    /// Fetch content from the origin
+    async fn get_origin(&self, root_cid: Cid) -> Result<()> {
+        info!("Fetching cid {root_cid} from origin (ipfs)");
+        let pending = self.pending_requests.clone();
+        let (tx, mut rx) = unbounded_channel();
+        match self.pending_requests.write().await.entry(root_cid) {
+            Entry::Occupied(mut e) => {
+                // there is a concurrent request for this cid, just wait for the first one and return
+                e.get_mut().push(tx);
+                return rx
+                    .recv()
+                    .await
+                    .ok_or_else(|| anyhow!("Failed to receive status from channel"))?;
+            }
+            Entry::Vacant(e) => {
+                e.insert(vec![tx]);
+            }
+        }
+
+        // we are the first concurrent request for this cid
+        let client = self.client.clone();
+        let req = RequestBuilder::new(
+            Method::Get,
+            format!(
+                "https://{}/ipfs/{root_cid}",
+                self.origin_config.ipfs_gateway
+            )
+            .parse()?,
+        )
+        .header("Accept", "application/vnd.ipld.car")
+        .build();
+
+        let store = self.store.db.clone();
+        spawn(async move {
+            // send the request
+            let res = async {
+                let res = client.send(req).await.map_err(|e| {
+                    format!("Error getting content for cid {root_cid} from origin: {e}")
+                })?;
+
+                let mut car = CarReader::new(res).await.map_err(|e| {
+                    format!("Error reading car file for cid {root_cid} from origin: {e}")
+                })?;
+
+                if car.header.roots.iter().any(|cid| *cid == root_cid) {
```

Comment:
why aren't we using `load_car` directly?

Reply:
also could we use the internal `put_file` method

Reply:
I wanted to verify the cid is present in headers before caching the content

Reply:
How does that prevent from using load car directly tho

Reply:
I tried using CarReader and passing load_car(store, car.reader) but for some reason it's erroring when the CarReaders are nested

Reply:
Interesting, you don't think using existing code like `put_file` or `put_car` is a good option. I am wondering why?

Reply:
We have the data, owned and in memory already, why write to disk -> read from disk in put_file -> write to store in load_car, and finally read from store/traverse dag to return the data from `get`/`get_data`. And that's not even including if we want to ensure the root cid is present

The flow now is parse car -> verify root is present with CarReader -> insert directly into blockstore -> final read/traveral to return the data

Reply:
In terms of put_car, maybe that could be used but I still think we should check to ensure the root cid is present

Reply:
actually looking into it the previous implementation on current main is not properly handling this either for bitswap when cid is missing

Reply:
Put file calls put car which in turn calls load car. 

So if you are using load_car then put_car is the same

Reply:
We have to store the content from origin right? Because it's a cache miss

Reply:
> ... we still need to emit the provider and network command to trigger advertisement and content replication

Yes, and that's important for indexing. Another data point that is important to the gateway is the size of the file so we need to handle provide that. Please see #221.

Reply:
cc @scott-dn

Reply:
resolving this, latest convo here on the extra verification https://github.com/fleek-network/ursa/pull/278#discussion_r1069939919

and provider and network is correct now